ghga_validator/plugins/unique_validation.py

```diff
@@ -0,0 +1,104 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Plugin for LinkML JSON Validator used for validating the non inline references"""
+
+from collections import defaultdict
+from typing import Dict, List
+
+from linkml_runtime.utils.schemaview import SchemaView
+from linkml_validator.models import SeverityEnum, ValidationMessage, ValidationResult
+from linkml_validator.plugins.base import BasePlugin
+
+from ghga_validator.linkml.object_iterator import ObjectIterator
+from ghga_validator.utils import path_as_string
+
+
+# pylint: disable=too-many-locals
+class UniqueValidationPlugin(BasePlugin):
+    """
+    Plugin to check whether the fields defined as identifier/unique key
+    are unique for a class.
+
+    Args:
+        schema: Path or URL to schema YAML
+        kwargs: Additional arguments that are used to instantiate the plugin
+
+    """
+
+    NAME = "UniqueValidationPlugin"
+
+    def __init__(self, schema: str) -> None:
+        super().__init__(schema)
+        self.schemaview = SchemaView(schema)
+
+    def process(self, obj: Dict, **kwargs) -> ValidationResult:
+        """
+        Perform validation on an object.
+
+        Args:
+            obj: The object to validate
+            kwargs: Additional arguments that are used for processing
+
+        Returns:
+            ValidationResult: A validation result that describes the outcome of validation
+
+        """
+        target_class = kwargs["target_class"]
+
+        messages = self.validate_unique_fields(obj, target_class)
+        valid = len(messages) == 0
+
+        result = ValidationResult(
+            plugin_name=self.NAME, valid=valid, validation_messages=messages
+        )
+        return result
+
+    def validate_unique_fields(
+        self,
+        object_to_validate: Dict,
+        target_class: str,
+    ) -> List[ValidationMessage]:
+        """
+        Validate non inlined reference fields in a JSON object
+
+        Args:
+            object_to_validate: input JSON object
+            target_class: parent class in the schema
+
+        Returns:
+            SlotDefinition: class definition
+
+        """
+        messages = []
+
+        all_ids = defaultdict(list)  # type: ignore
+
+        for class_name, identifier, data, path in ObjectIterator(
+            self.schemaview, object_to_validate, target_class
+        ):
+            if identifier in all_ids[class_name]:
+                id_slot = self.schemaview.get_identifier_slot(class_name)
+                message = ValidationMessage(
+                    severity=SeverityEnum.error,
+                    message="Duplicate value for unique attribute "
+                    + f"{class_name}({id_slot.name}): {identifier}",
+                    field=f"{path_as_string(path)}",
+                    value=data,
+                )
+                messages.append(message)
+            else:
+                all_ids[class_name].append(identifier)
```

Comment:
```suggestion
        seen_ids: Dict[Tuple, List] = {}

        for class_name, identifier, data, path in ObjectIterator(
            self.schemaview, object_to_validate, target_class
        ):
            id_slot = self.schemaview.get_identifier_slot(class_name)
            id_slot_name = id_slot.name if id_slot is not None else "UNKNOWN"

            if (class_name, identifier) in seen_ids:
                previous_path = seen_ids[class_name, identifier]
                message = ValidationMessage(
                    severity=SeverityEnum.error,
                    message="Duplicate value for identifier, "
                    + f"same value used at {path_as_string(previous_path)}.",
                    field=f"{path_as_string(path)}.{id_slot_name}",
                    value=data[id_slot_name],
                )
                messages.append(message)
            else:
                seen_ids[class_name, identifier] = path + [id_slot_name]
```

I think it's more intuitive to actually use the path including the identifier field since that's where the "wrong" value is set. Also pointing to the alternative path using the same ID is helpful.